### PR TITLE
188 add request repeater

### LINKED ISSUE #188: Add request repeater for 5xx errors
### Description
It should automatically repeat on particular 5xx errors like [Service Unavailable](https://dev.commercetools.com/http-api-errors.html#503-service-unavailable)

### Expected Behavior
Repeat on 5xx API errors and accept optionally configured list of errors/codes for which it should retry/repeat.

### Current Behavior
Throws 503 error without repeating.

### Context
It would allow reliable interaction with API.

### Possible Solution
https://github.com/sphereio/sphere-node-sdk/pull/190

### PR DESCRIPTION
**Pull request by hisabimbola**
#### Summary

Fixes #188 
#### Todo

- Tests
    - [ ] Unit
    - [ ] Integration
    - [ ] Acceptance
- [ ] Documentation

### CODE REVIEW

File: docs/sdk/api/sdkMiddlewareHttp.md [hunk 1]
```diff
@@ -24,6 +24,12 @@ Creates a [middleware](/sdk/Glossary.md#middleware) to handle HTTP requests for
 2. `includeResponseHeaders` *(Boolean)*: flag whether to include the response headers in the response, if omitted headers is omitted from response
 3. `includeOriginalRequest` *(Boolean)*: flag whether to include the original request sent in the response. Can be useful if you want to see the final request being sent.
 4. `maskSensitiveHeaderData` *(Boolean)*: flag to mask sensitie data in the header. e.g. Authorization token
+5. `enableRetry` *(Boolean)*: flag to enable retry on network and 500 requests
```
Comment (wizzy25) [inline docs/sdk/api/sdkMiddlewareHttp.md]: Nitpick: `500 request` of `500 error`?

File: packages/sdk-middleware-http/src/http.js [hunk 2]
```diff
@@ -41,68 +47,82 @@ export default function createHttpMiddleware (
...
+          if (enableRetry)
+            if (retryCount < maxRetries) {
+              retryCount += 1
+              setTimeout(executeFetch, retryDelay)
+              return
+            }
+          const error = new NetworkError(
+            e.message, { originalRequest: request },
+          )
+          next(request, { ...response, error, statusCode: 0 })
```
Comment (wizzy25) [inline packages/sdk-middleware-http/src/http.js]: Why do you set the `statusCode` to 0?

File: packages/sdk-middleware-http/test/http.spec.js [hunk 3]
```diff
@@ -239,6 +238,42 @@ describe('Http', () => {
     }),
   )
 
+  it('handle failed response (network error)', () =>
```
Comment (butenkor) [inline packages/sdk-middleware-http/test/http.spec.js]: i would add more tests:
- that repeater is not executed on 400/409/or other not 5xx errors
- that it rejects after X failed retries with last/most recent error response
- that it resolves after X successful retry

File: packages/sdk-middleware-http/src/http.js [hunk 4]
```diff
@@ -41,68 +47,82 @@ export default function createHttpMiddleware (
...
-      (e: Error) => {
-        const error = new NetworkError(e.message, { originalRequest: request })
-        next(request, { ...response, error, statusCode: 0 })
-      },
-    )
+        },
+        // We know that this is a "network" error thrown by the `fetch` library
+        (e: Error) => {
+          if (enableRetry)
+            if (retryCount < maxRetries) {
```
Comment (butenkor) [inline packages/sdk-middleware-http/src/http.js]: It would be very helpful when debugging logs to know that logged requests has been retried X amount of times. On last attempt SDK should notify the caller that it tried X amount of times and failed with most recent error message: https://github.com/sphereio/sphere-node-utils/blob/master/src/coffee/helpers/repeater.coffee#L61

File: docs/sdk/api/sdkMiddlewareHttp.md (hunk 1 above)
Comment (adnasa) [inline docs/sdk/api/sdkMiddlewareHttp.md]: more like `500 response`?

File: package.json [hunk 5]
```diff
@@ -58,6 +58,11 @@
     "eslint-plugin-react": "7.0.1",
     "flow-bin": "0.47.0",
     "gitbook-cli": "2.3.0",
+    "gitbook-plugin-anchorjs": "^1.1.1",
+    "gitbook-plugin-edit-link": "^2.0.2",
+    "gitbook-plugin-ga": "^1.0.1",
+    "gitbook-plugin-github": "^2.0.0",
+    "gitbook-plugin-prism": "^2.2.0",
```
Comment (adnasa) [inline package.json]: Is this really related to the scope of this feature?

File: docs/sdk/api/sdkMiddlewareHttp.md [hunk 6]
```diff
@@ -24,6 +24,12 @@ Creates a [middleware](/sdk/Glossary.md#middleware) to handle HTTP requests for
...
+6. `maxRetries` *(Number)*: number of times to retry the request before failing the request. (Default: 50)
+7. `retryDelay` *(Number)*: amount of milliseconds to wait before retrying the next request. (Default: 200)
+
+#### Retrying requests
+This modules have a retrying ability incase of network failures or 503 response errors. To enable this behavior, pass the `enableRetry` flag in the options and also set the maximum number of retries (`maxRetries`) and amount of milliseconds to wait before retrying a request (`retryDelay`).
```
Comment (junajan) [inline docs/sdk/api/sdkMiddlewareHttp.md]: `This modules have a retrying ability incase of network failures` -> `This module have a retrying ability in case of network failures`

File: packages/sdk-middleware-http/src/http.js [hunk 7]
```diff
@@ -41,68 +47,82 @@ export default function createHttpMiddleware (
...
+              retryCount += 1
+              setTimeout(executeFetch, retryDelay)
```
Comment (hisabimbola) [inline packages/sdk-middleware-http/src/http.js]: Added exponential backoff. Thanks for highlighting that.

and amidst several reasons, the old repeater module is not optimise for the browser

File: packages/sdk-middleware-http/test/http.spec.js (hunk 3 above)
Comment (hisabimbola) [inline packages/sdk-middleware-http/test/http.spec.js]: Added more relevant test

File: docs/sdk/api/sdkMiddlewareHttp.md (hunk 1 above)
Comment (butenkor) [inline docs/sdk/api/sdkMiddlewareHttp.md]: Most SDKs have it enabled by default. See Amazon or Paypal SDKs for example. As a browser user i would rather wait a bit longer (2-3 retries) and see successful processing instead of embarrassing 5xx error. For browser you can use shorter and fixed time interval.

File: packages/sdk-middleware-http/src/http.js [hunk 8]
```diff
@@ -41,68 +47,82 @@ export default function createHttpMiddleware (
...
+          // Server responded with an error. Try to parse it as JSON, then
+          // return a proper error type with all necessary meta information.
+          res.text()
+          .then((text: any) => {
+            // Try to parse the error response as JSON
+            let parsed
+            try {
+              parsed = JSON.parse(text)
+            } catch (error) {
+              /* noop */
```
Comment (butenkor) [inline packages/sdk-middleware-http/src/http.js]: Wrong commit link?

File: packages/sdk-middleware-http/src/http.js [hunk 9]
```diff
@@ -41,68 +47,82 @@ export default function createHttpMiddleware (
...
-        })
-      },
-      // We know that this is a "network" error thrown by the `fetch` library
-      (e: Error) => {
-        const error = new NetworkError(e.message, { originalRequest: request })
-        next(request, { ...response, error, statusCode: 0 })
-      },
-    )
+        },
+        // We know that this is a "network" error thrown by the `fetch` library
```
Comment (butenkor) [inline packages/sdk-middleware-http/src/http.js]: Will it retry on all listed error cases (as it currently does in old SDK)? If not how can we ensure it or make configurable?

File: packages/sdk-middleware-http/src/http.js (hunk 4 above)
Comment (butenkor) [inline packages/sdk-middleware-http/src/http.js]: Yes it is needed for logs in order to create dashboards, alerts, notifications etc. It should be possible for the user to know that rejected request has been retried.

File: packages/sdk-middleware-http/test/http.spec.js (hunk 3 above)
Comment (butenkor) [inline packages/sdk-middleware-http/test/http.spec.js]: Could you please also add a test for last case?

File: packages/sdk-middleware-http/src/http.js [hunk 10]
```diff
@@ -41,68 +47,82 @@ export default function createHttpMiddleware (
...
       },
     )
-    fetch(requestObj)
-    .then(
-      (res: Response) => {
-        if (res.ok) {
-          res.json()
-          .then((result: Object) => {
-            const parsedResponse: Object = {
+    let retryCount = 0
```
Comment (hisabimbola) [inline packages/sdk-middleware-http/src/http.js]: Yeah, just not sure it's better than what we have now

File: packages/sdk-middleware-http/src/http.js (hunk 8 above)
Comment (hisabimbola) [inline packages/sdk-middleware-http/src/http.js]: Yup, still working on it, I will add it shortly

File: packages/sdk-middleware-http/src/http.js (hunk 9 above)
Comment (hisabimbola) [inline packages/sdk-middleware-http/src/http.js]: Yeah, it _should_ retry on all listed errors.

Basically anything that is not a response from the API, will go to the catch block, according to the spec

File: packages/sdk-middleware-http/src/http.js (hunk 4 above)
Comment (hisabimbola) [inline packages/sdk-middleware-http/src/http.js]: Yeah, we can log it.

File: packages/sdk-middleware-http/src/http.js (hunk 9 above)
Comment (butenkor) [inline packages/sdk-middleware-http/src/http.js]: Will it retry on 5xx errors from API also?

File: packages/sdk-middleware-http/src/http.js (hunk 4 above)
Comment (butenkor) [inline packages/sdk-middleware-http/src/http.js]: The idea is rather sdk does not log anything but let the caller log it if required. Sdk should just throw custom error with custom message. Custom error should always supply root error/message

File: docs/sdk/api/sdkMiddlewareHttp.md [hunk 11]
```diff
@@ -25,12 +25,17 @@ Creates a [middleware](/sdk/Glossary.md#middleware) to handle HTTP requests for
 3. `includeOriginalRequest` *(Boolean)*: flag whether to include the original request sent in the response. Can be useful if you want to see the final request being sent.
 4. `maskSensitiveHeaderData` *(Boolean)*: flag to mask sensitie data in the header. e.g. Authorization token
 5. `enableRetry` *(Boolean)*: flag to enable retry on network errors and `500` response. (Default: false)
-6. `maxRetries` *(Number)*: number of times to retry the request before failing the request. (Default: 50)
-7. `retryDelay` *(Number)*: amount of milliseconds to wait before retrying the next request. (Default: 200)
+6. `retryConfig` *(Object)*: Field required in the object listed below
+  1. `maxRetries` *(Number)*: number of times to retry the request before failing the request. (Default: 50)
+  2. `retryDelay` *(Number)*: amount of milliseconds to wait before retrying the next request. (Default: 200)
+  3. `backoff` *(Boolean)*: activates exponential backoff. Recommended to prevent spamming of the server. (Default: true)
+  4. `maxDelay` *(Number)*: The maximum duration (milliseconds) to wait before retrying, useful if the delay time grew exponentially more than reasonable
```
Comment (hisabimbola) [inline docs/sdk/api/sdkMiddlewareHttp.md]: Useful if used in browser, then you set a maximum time for requests. then you don't have to manually go and calculate what the maximum delay will be based on the `retryDelay` and `maxRetries`

File: packages/sdk-middleware-http/src/http.js (hunk 8 above)
Comment (hisabimbola) [inline packages/sdk-middleware-http/src/http.js]: Fixed here ff24e781c75e790a06d99d23cbea9f8f95796db9

File: packages/sdk-middleware-http/src/http.js (hunk 9 above)
Comment (hisabimbola) [inline packages/sdk-middleware-http/src/http.js]: Yes, it will.

File: docs/sdk/api/sdkMiddlewareHttp.md (hunk 11 above)
Comment (adnasa) [inline docs/sdk/api/sdkMiddlewareHttp.md]: Can you prove this, @butenkor ?

---

I concur on @hisabimbola direction on this. and combining `maxRetries` with `retryDelay` doesn't solve the purpose of what `maxDelay` is written for until proven otherwise.
Even if that is the case, we essentially developers to hack around the system of this feature

File: packages/sdk-middleware-http/src/http.js (hunk 4 above)
Comment (hisabimbola) [inline packages/sdk-middleware-http/src/http.js]: Added here 2d49400e3c96c14a4bcf516ddea4049d28de618e

**Comment by hisabimbola** [comment]
Merging this PR now, we can always improve on it.

File: docs/sdk/api/sdkMiddlewareHttp.md (hunk 11 above)
Comment (butenkor) [inline docs/sdk/api/sdkMiddlewareHttp.md]: ok i give up :)

**Comment by butenkor** [comment]
@hisabimbola Thanks!

File: docs/sdk/api/sdkMiddlewareHttp.md (hunk 11 above)
Comment (adnasa) [inline docs/sdk/api/sdkMiddlewareHttp.md]: @butenkor 😝
